### PR TITLE
refactor: update the client retry by the pkg retryablehttp

### LINKED ISSUE #2749: Why retry so many times for the builder?
**Fission/Kubernetes version**

<pre>
$ fission version
$ kubectl version
</pre>

**Kubernetes platform (e.g. Google Kubernetes Engine)**

**Describe the bug**

https://github.com/fission/fission/commit/23942fdf7db917eae3052e8c4b7e356c0c189b63#diff-16303cbae6f1706622e3c61a3329d7989c89bc30dab964865c8d11852c1807ed

There is a logic of 20 retries, and there is no distinction between network errors and build errors. that's not useful for the build error, And A large number of useless retries take up a lot of memory, they crowd out the resources of normal build tasks.

**To Reproduce**

A limit memory builder with a normal function and a build error function

**Expected result**

Only retry for the network error. Or the HTTP status code > 500

**Actual result**
Any not 200 status code will be retried

**Screenshots/Dump file**

<pre>
$ fission support dump
</pre>

**Additional context**

### PR DESCRIPTION
**Pull request by saltbo**
## Description

The old retry policy is too simple, we can use the pkg retryablehttp to replace it

## Which issue(s) this PR fixes:

Fixes #2749

## Checklist:

- [x] I ran tests as well as code linting locally to verify my changes. 
- [ ] I have done manual verification of my changes, changes working as expected.
- [ ] I have added new tests to cover my changes.
- [x] My changes follow contributing guidelines of Fission.
- [x] I have signed all of my commits.

### CODE REVIEW

**Review by sanketsudake — APPROVED** [review]

File: pkg/builder/client/client.go [hunk 1]
```diff
@@ -39,12 +38,13 @@ type (
...
 		logger     *zap.Logger
 		url        string
-		httpClient *http.Client
+		httpClient *retryablehttp.Client
 	}
 )
 
 func MakeClient(logger *zap.Logger, builderUrl string) *Client {
-	hc := &http.Client{Transport: otelhttp.NewTransport(http.DefaultTransport)}
+	hc := retryablehttp.NewClient()
```
Comment (shubham-bansal96) [inline pkg/builder/client/client.go:46]: You should define maxRetries here as well.
hc.RetryMax = 20

**Review by shubham-bansal96 — CHANGES_REQUESTED** [review]

File: pkg/builder/client/client.go (hunk 1 above)
Comment (sanketsudake) [inline pkg/builder/client/client.go:46]: Default max retry time is 30 seconds where as min is 1 in library.
Max tries default is 4.
https://github.com/hashicorp/go-retryablehttp/blob/7f638666931a119b8d7a1239e44da11db0d446a4/client.go#L47

We need to see if we are honoring current max retry time.

Comment (saltbo) [reply pkg/builder/client/client.go:46]: > However 20 retries is too much we and reduce retries and increase wait duration.

Yes, I agree. The original intention of this PR is to change the number of retries and retry strategy.

The current retry policy is from the PR Istio integration (https://github.com/fission/fission/pull/421), I think it wants to retry only for the network error, not the build error.  short time with 20 retries for the network error may be suitable, But not suitable for the build error.

Comment (shubham-bansal96) [reply pkg/builder/client/client.go:46]: Make sense. As per current logic written in code, max retry time won't not more than 2 second.
So i think we can go ahead with these changes without specifying `hc.RetryMax = 20`

**Review by shubham-bansal96 — APPROVED** [review]

**Pull request merged**